### PR TITLE
quiche: update QUICHE to commit 5dd7a030209f9a6b5043bebd8ac3ee54f18d1d08

### PR DESCRIPTION
**Pull request by danzh2010**
Signed-off-by: Dan Zhang <danzh@google.com>

Commit Message: Modify QUIC_BUG implementation to log with rate limit like ENVOY_BUG in release mode. Fix a QuicMemSliceSpanImpl life time issue which was exposed by new QUICHE change

Risk Level: low
Testing: added unit tests for QUIC_BUG and QuicMemSliceSpanImpl

### CODE REVIEW

**Comment by moderation** [comment]
You have some CI issues but...

/lgtm deps

**Comment by danzh2010** [comment]
It's ready for review now. PTAL!
Coverage test failure is caused by ProxyProtoFilterChainMatchIntegrationTest.MismatchDirectSourceButMatchSource/IPv4 which is irrelevant to this PR.

File: source/common/quic/platform/quic_mem_slice_span_impl.h
Comment (wu-bin) [inline source/common/quic/platform/quic_mem_slice_span_impl.h:64]: Can you add some comments to explain the relationship of buffer_ and mem_slice_?  (It seems we only need one of them to be !nullptr at a time, which is easier to understand)

File: source/common/quic/platform/quiche_bug_tracker_impl.h
Comment (wu-bin) [inline source/common/quic/platform/quiche_bug_tracker_impl.h:30]: nit: Clarify (in comments) it is test only, and only one instance can exists at a time.

File: source/common/quic/platform/quiche_bug_tracker_impl.cc
Comment (wu-bin) [inline source/common/quic/platform/quiche_bug_tracker_impl.cc:37]: nit: remove .data()

File: test/common/quic/platform/quic_platform_test.cc
Comment (wu-bin) [inline test/common/quic/platform/quic_platform_test.cc:124]: "should not be logged".  If there's a bug counter it should still be incremented.

File: source/common/quic/platform/quic_mem_slice_span_impl.h
Comment (danzh2010) [reply source/common/quic/platform/quic_mem_slice_span_impl.h:64]: done. buffer_ always points to the actually memory, but mem_slice_ is only used in the constructor which takes QuicMemSlice.

File: source/common/quic/platform/quiche_bug_tracker_impl.h
Comment (danzh2010) [reply source/common/quic/platform/quiche_bug_tracker_impl.h:30]: done. But why only one instance can exists at a time?

File: source/common/quic/platform/quiche_bug_tracker_impl.cc
Comment (danzh2010) [reply source/common/quic/platform/quiche_bug_tracker_impl.cc:37]: invokeEnvoyBugFailureRecordActionForEnvoyBugMacroUseOnly() take a char*

File: source/common/quic/platform/quiche_bug_tracker_impl.h
Comment (wu-bin) [reply source/common/quic/platform/quiche_bug_tracker_impl.h:30]: I don't see comment added?

Suppose two threads, each has a ScopedDisableExitOnQuicheBug instance, called S1 and S2, and S1 and S2's lifetime overlaps like:

T=0:   S1 construct
T=1:                                S2 construct
T=2:   S1 destruct
T=3:                               S2 destruct

The final value for ScopedDisableExitOnQuicheBug will be true.

Comment (danzh2010) [reply source/common/quic/platform/quiche_bug_tracker_impl.h:30]: Oh, you mean multiple thread. Add comment for real now.

Comment (wu-bin) [reply source/common/quic/platform/quiche_bug_tracker_impl.h:30]: Yes multi thread is the main problem. But that sequence is problematic in single thread too.

Still not seeing comments. LGTM assuming you'll add one:)

**Review by wu-bin — APPROVED** [review]
Thanks Dan. LGTM but please add a comment for ScopedDisableExitOnQuicheBug.

**Comment by alyssawilk** [comment]
(tagging as waiting on a new hash per internal discussion)

**Comment by danzh2010** [comment]
> (tagging as waiting on a new hash per internal discussion)

Updated to later commit. PTAL

File: bazel/repository_locations.bzl
Comment (moderation) [inline bazel/repository_locations.bzl:816]: I _think_ the date for this commit is `2021-07-16` as per https://quiche.googlesource.com/quiche/+/5dd7a030209f9a6b5043bebd8ac3ee54f18d1d08 but it's a little difficult to navigate `googlesource.com`. As this isn't a Github repo I don't think we check this programatically

**Review by alyssawilk — APPROVED** [review]
LGTM modulo outstanding comments

File: bazel/repository_locations.bzl
Comment (danzh2010) [reply bazel/repository_locations.bzl]: fixed

**Comment by alyssawilk** [comment]
@moderation I think this needs one more LGTM?

**Pull request merged**
